test/PositionContext.test.ts

```diff
@@ -1084,97 +1099,154 @@ suite("PositionContext", () => {
+                                    (62 <= i && i <= 63) ? [propertyCompletion("name", i, 0)] :
                                         []);
+                }
+
+                for (let i = 0; i <= 69; ++i) {
+                    completionItemsTest(`{ "variables": { "a": { "name": "A" } }, "b": "[variables('a').na]" }`, i,
+                        (i === 48) ? allCompletions(48, 0) :
+                            (49 <= i && i <= 57) ? [variablesCompletion(48, 9)] :
+                                (59 <= i && i <= 60) ? [variableCompletion("a", 58, 4)] :
+                                    (62 <= i && i <= 65) ? [propertyCompletion("name", 63, 2)] :
+                                        []);
+                }
+
+                for (let i = 0; i <= 69; ++i) {
+                    completionItemsTest(`{ "variables": { "a": { "name": "A" } }, "b": "[variables('a').ab]" }`, i,
+                        (i === 48) ? allCompletions(48, 0) :
+                            (49 <= i && i <= 57) ? [variablesCompletion(48, 9)] :
+                                (59 <= i && i <= 60) ? [variableCompletion("a", 58, 4)] :
+                                    (62 <= i && i <= 63) ? [propertyCompletion("name", 63, 2)] :
+                                        []);
+                }
+
+                for (let i = 0; i <= 78; ++i) {
+                    completionItemsTest(`{ "variables": { "a": { "bb": { "cc": 200 } } }, "b": "[variables('a').bb.]" }`, i,
+                        (i === 56) ? allCompletions(56, 0) :
+                            (57 <= i && i <= 65) ? [variablesCompletion(56, 9)] :
+                                (67 <= i && i <= 68) ? [variableCompletion("a", 66, 4)] :
+                                    (70 <= i && i <= 73) ? [propertyCompletion("bb", 71, 2)] :
+                                        (i === 74) ? [propertyCompletion("cc", 74, 0)] :
+                                            []);
+                }
+
+            });
+
+            function getDocumentAndMarkers(document: object | string): { documentText: string; tokens: number[] } {
+                let tokens: number[] = [];
+                document = typeof document === "string" ? document : JSON.stringify(document);
+
+                // tslint:disable-next-line:no-constant-condition
+                while (true) {
+                    let tokenPos = document.indexOf("!");
+                    if (tokenPos < 0) {
+                        break;
+                    }
+                    tokens.push(tokenPos);
+                    document = document.slice(0, tokenPos) + document.slice(tokenPos + 1);
+                }
+
+                return {
+                    documentText: document,
+                    tokens
+                };
             }
+
+            suite("Parameter defaultValue deep completion for objects", () => {
+                let { documentText, tokens } = getDocumentAndMarkers({
+                    parameters: {
```

Comment:
So it's not obvious to me that I have to define a defaultValue to get completion for this. I also might want completion for properties not included in the defaultValue. Can I specify the schema for parameters of type object? I'm talking mainly about the "properties" object like [in this example](http://json-schema.org/learn/miscellaneous-examples.html):
```
{
  "$id": "https://example.com/person.schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Person",
  "type": "object",
  "properties": {
    "firstName": {
      "type": "string",
      "description": "The person's first name."
    },
    "lastName": {
      "type": "string",
      "description": "The person's last name."
    },
    "age": {
      "description": "Age in years which must be equal to or greater than zero.",
      "type": "integer",
      "minimum": 0
    }
  }
}
```

In other words, I would expect "firstName", "lastName", and "age" to be the completion entries instead of whatever is in a defaultValue.

Reply:
No, you can't currently do that in ARM templates, but good question.